pkg/agent/controller/bgp/controller.go

```diff
@@ -571,56 +588,95 @@ func (c *Controller) addServiceRoutes(advertisement *v1alpha1.ServiceAdvertiseme
 		if ipTypes.Has(v1alpha1.ServiceIPTypeClusterIP) {
 			if internalLocal && hasLocalEndpoints || !internalLocal {
 				for _, clusterIP := range svc.Spec.ClusterIPs {
-					serviceIPs = append(serviceIPs, clusterIP)
+					if c.enabledIPv4 && utilnet.IsIPv4String(clusterIP) {
+						allRoutes[bgp.Route{Prefix: clusterIP + ipv4Suffix}] = RouteMetadata{
+							Type:      ServiceClusterIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
+					if c.enabledIPv6 && utilnet.IsIPv6String(clusterIP) {
+						allRoutes[bgp.Route{Prefix: clusterIP + ipv6Suffix}] = RouteMetadata{
+							Type:      ServiceClusterIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
 				}
 			}
 		}
 		if ipTypes.Has(v1alpha1.ServiceIPTypeExternalIP) {
 			if externalLocal && hasLocalEndpoints || !externalLocal {
 				for _, externalIP := range svc.Spec.ExternalIPs {
-					serviceIPs = append(serviceIPs, externalIP)
+					if c.enabledIPv4 && utilnet.IsIPv4String(externalIP) {
+						allRoutes[bgp.Route{Prefix: externalIP + ipv4Suffix}] = RouteMetadata{
+							Type:      ServiceExternalIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
+					if c.enabledIPv6 && utilnet.IsIPv6String(externalIP) {
+						allRoutes[bgp.Route{Prefix: externalIP + ipv6Suffix}] = RouteMetadata{
+							Type:      ServiceExternalIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
 				}
 			}
 		}
 		if ipTypes.Has(v1alpha1.ServiceIPTypeLoadBalancerIP) && svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
 			if externalLocal && hasLocalEndpoints || !externalLocal {
-				serviceIPs = append(serviceIPs, getIngressIPs(svc)...)
+				loadBalancerIPs := getIngressIPs(svc)
+				for _, loadBalancerIP := range loadBalancerIPs {
+					if c.enabledIPv4 && utilnet.IsIPv4String(loadBalancerIP) {
+						allRoutes[bgp.Route{Prefix: loadBalancerIP + ipv4Suffix}] = RouteMetadata{
+							Type:      ServiceLoadBalancerIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
+					if c.enabledIPv6 && utilnet.IsIPv6String(loadBalancerIP) {
+						allRoutes[bgp.Route{Prefix: loadBalancerIP + ipv6Suffix}] = RouteMetadata{
+							Type:      ServiceLoadBalancerIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
```

Comment:
to avoid the code repetition, maybe define a helper closure:
```go
addRoutes := func(svc string, ip string, routeType AdvertisedRouteType){ ... }
```